### PR TITLE
chore: drop python 3.8 support and upgrade ETH dependencies

### PR DESCRIPTION
**Pull request by antazoey**
### What I did

fixes: #

### Checklist

- [ ] Passes all linting checks (pre-commit and CI jobs)
- [ ] New test cases have been added and are passing
- [ ] Documentation has been updated
- [ ] PR title follows [Conventional Commit](https://www.conventionalcommits.org/en/v1.0.0/) standard (will be automatically included in the changelog)

### CODE REVIEW

**Comment by fubuloubu** [comment]
related to the other discussion, I wonder if these lower-level packages than ape should stay 3.8 a little bit longer, enough time for those using ape to upgrade to 0.8 where py 3.8 is dropped

then we drop this closer to the deprecation date (maybe when we go v1.0 with ape)

**Comment by antazoey** [comment]
> related to the other discussion, I wonder if these lower-level packages than ape should stay 3.8 a little bit longer, enough time for those using ape to upgrade to 0.8 where py 3.8 is dropped
> 
> then we drop this closer to the deprecation date (maybe when we go v1.0 with ape)

I am going back and forth. The main thing is I need the eth dependencies upgraded and I can't do that without also it being breaking (think Hexbytes change). So I need breaking changes releases for all these lower levels packages to even upgrade ape. I guess a chicken and egg problem.

File: tests/test_sourcemap.py [hunk 1]
```diff
@@ -1,4 +1,4 @@
-from typing import Iterator
+from collections.abc import Iterator
```
Comment (fubuloubu) [inline tests/test_sourcemap.py:1]: what's the story behind this change? I saw this somewhere else and didn't get why

File: setup.py [hunk 2]
```diff
@@ -76,14 +75,14 @@
     url="https://github.com/ApeWorX/ethpm-types",
     include_package_data=True,
     install_requires=[
-        "pydantic>=2.4.2,<3",
-        "eth-utils>=2.1.0,<3",
+        "pydantic>=2.6.4,<3",
+        "eth-utils>=4.1.0,<5",
```
Comment (fubuloubu) [inline setup.py:79]: In general, I wonder how wide our pins can be for these EF packages, I think it's best even to leave them open above (and keep the tighter pins in Ape if we need it)

File: ethpm_types/utils.py [hunk 3]
```diff
@@ -1,14 +1,15 @@
 import json
+from collections.abc import Sequence
 from enum import Enum
 from hashlib import md5, sha3_256, sha256
-from typing import Any, Dict, List, Optional, Sequence, Tuple
```
Comment (fubuloubu) [inline ethpm_types/utils.py:4]: Sequence is still being used here

File: ethpm_types/utils.py [hunk 4]
```diff
@@ -1,14 +1,15 @@
...
+from typing import Any, Optional
 
 from eth_pydantic_types import HexStr
 
 try:
     from typing import Annotated  # type: ignore
 except ImportError:
-    from typing_extensions import Annotated  # type: ignore
+    from typing import Annotated  # type: ignore
```
Comment (fubuloubu) [inline ethpm_types/utils.py:12]: can eliminate this and move `Annotated` above

File: ethpm_types/contract_type.py [hunk 5]
```diff
@@ -503,14 +504,16 @@ def _get_first_instance(self, _type: Type[ABI_SINGLETON_T]) -> Optional[ABI_SING
...
+    def _abi_identifiers(self) -> list[tuple[ABI_W_SELECTOR_T, str]]:
         def get_id(aitem: ABI_W_SELECTOR_T) -> str:
-            if isinstance(aitem, MethodABI) or isinstance(aitem, ErrorABI):
-                return HexBytes(self._selector_hash_fn(aitem.selector)[:4]).hex()
-            else:
-                return HexBytes(self._selector_hash_fn(aitem.selector)).hex()
+            bytes_val = (
+                self._selector_hash_fn(aitem.selector)[:4]
+                if isinstance(aitem, (MethodABI, ErrorABI))
+                else self._selector_hash_fn(aitem.selector)
```
Comment (fubuloubu) [inline ethpm_types/contract_type.py:512]: I think of this one as the weird one (basically for `EventABI` it only uses 32 bytes)

File: ethpm_types/contract_type.py [hunk 6]
```diff
@@ -1,5 +1,6 @@
+from collections.abc import Callable, Iterable
 from functools import cached_property, singledispatchmethod
-from typing import Callable, Dict, Iterable, List, Optional, Tuple, Type, TypeVar, Union, cast
```
Comment (fubuloubu) [inline ethpm_types/contract_type.py:2]: I think `Type` is still in use

File: ethpm_types/contract_type.py [hunk 7]
```diff
@@ -1,5 +1,6 @@
+from collections.abc import Callable, Iterable
```
Comment (fubuloubu) [inline ethpm_types/contract_type.py:1]: same here, is it suggested to import Callable from here instead as of some python version?

File: tests/test_sourcemap.py (hunk 1 above)
Comment (antazoey) [reply tests/test_sourcemap.py:1]: typing has aliases to collections.abc but theyre deprecated. Pre python 3.9, they were required

File: ethpm_types/contract_type.py (hunk 6 above)
Comment (antazoey) [reply ethpm_types/contract_type.py:2]: You use lowercase `type` now , no import needed!

File: ethpm_types/contract_type.py (hunk 7 above)
Comment (antazoey) [reply ethpm_types/contract_type.py:1]: This is correct. Basically anything that is an alias in typing is now deprecated.

Comment (fubuloubu) [reply ethpm_types/contract_type.py:1]: ahhhh, okay I get it now

**Pull request merged**